libdocker/kube_docker_client.go

```diff
@@ -368,9 +368,9 @@ func (p *progressReporter) start() {
 		for {
 			select {
 			case <-ticker.C:
-				progress, timestamp := p.progress.get()
-				// If there is no progress for p.imagePullProgressDeadline, cancel the operation.
-				if time.Since(timestamp) > p.imagePullProgressDeadline {
+				status, progress, timestamp := p.progress.get()
+				// If there is no progress for p.imagePullProgressDeadline in 'Downloading' phase, cancel the operation.
+				if status != "Extracting" && time.Since(timestamp) > p.imagePullProgressDeadline {
```

Comment:
Come to think of it, there is a logical race here. [The status of the last progress update will be "Pull complete"](https://github.com/moby/moby/blob/34c3569768eb609ec1b26a73dd6803fcc9683b1c/distribution/xfer/download.go#L367), so if the ticker happens to tick at just the right moment the context will be canceled and a spurious error will be logged. While we could in theory enumerate all the possible statuses which should suppress cancelation, that would be quite brittle to daemon changes. What we can do instead is disable cancelation when the status becomes `Extracting` _and latch it disabled_ for the remainder of the operation.
```go
		downloaded := false
		for {
			select {
			case <-ticker.C:
				status, progress, timestamp := p.progress.get()
				// If there is no progress for p.imagePullProgressDeadline in 'Downloading' phase, cancel the operation.
				if status == "Extracting" {
						downloaded = true
				}
				if !downloaded && time.Since(timestamp) > p.imagePullProgressDeadline {
						// ...
```

Reply:
Thanks for the advice. Resolved in new commit [44b300b](https://github.com/Mirantis/cri-dockerd/pull/376/commits/44b300bfbea68e72cddc46224e64ceb553795217)